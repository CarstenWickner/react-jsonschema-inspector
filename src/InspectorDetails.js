/* eslint-disable quote-props */
import PropTypes from "prop-types";
import React, { PureComponent } from "react";

import InspectorDetailsForm from "./InspectorDetailsForm";
import JsonSchemaPropType from "./JsonSchemaPropType";
import { getFieldValue, getPropertyParentSchemas, isDefined } from "./utils";

class InspectorDetails extends PureComponent {
    static isSelectionRequiredInParent(columnData, selectionColumnIndex, refTargets) {
        if (selectionColumnIndex < 1) {
            // no parent to define any required properties
            return false;
        }
        const parentColumn = columnData[selectionColumnIndex - 1];
        const parentSchema = parentColumn.items[parentColumn.selectedItem];
        const targetItem = columnData[selectionColumnIndex].selectedItem;
        const schemaList = getPropertyParentSchemas(parentSchema, refTargets);
        return schemaList.some(part => isDefined(part.required) && part.required.includes(targetItem));
    }

    static collectFormFields(itemSchema, refTargets, columnData, selectionColumnIndex) {
        const formFields = [];
        const addFormField = (labelText, rowValue) => {
            if (isDefined(rowValue)) {
                formFields.push({ labelText, rowValue });
            }
        };
        const getValue = fieldName => getFieldValue(itemSchema, fieldName, refTargets);

        addFormField("Title", getValue("title"));
        addFormField("Description", getValue("description"));
        addFormField("Required",
            InspectorDetails.isSelectionRequiredInParent(columnData, selectionColumnIndex, refTargets) ? "Yes" : null);
        addFormField("Type", getValue("type"));
        addFormField("Constant Value", getValue("const"));
        addFormField("Possible Values", getValue("enum"));

        const minimum = getValue("minimum");
        const exclusiveMinimum = getValue("exclusiveMinimum");
        let minValue;
        if (isDefined(minimum)) {
            // according to JSON Schema Draft 4, "exclusiveMinimum" is a boolean and used in combination with "minimum"
            // if "exclusiveMinimum" is not defined, is is treated as "false", i.e. "minimum" is inclusive by default
            minValue = exclusiveMinimum ? `${minimum} (exclusive)` : `${minimum} (inclusive)`;
        } else {
            // according to JSON Schema Draft 6, "exclusiveMinimum" is a number and can be used instead of "minimum"
            minValue = isDefined(exclusiveMinimum) ? `${exclusiveMinimum} (exclusive)` : null;
        }
        addFormField("Min Value", minValue);

        const maximum = getValue("maximum");
        const exclusiveMaximum = getValue("exclusiveMaximum");
        let maxValue;
        if (isDefined(maximum)) {
            // according to JSON Schema Draft 4, "exclusiveMaximum" is a boolean and used in combination with "maximum"
            // if "exclusiveMaximum" is not defined, is is treated as "false", i.e. "maximum" is inclusive by default
            maxValue = exclusiveMaximum ? `${maximum} (exclusive)` : `${maximum} (inclusive)`;
        } else {
            // according to JSON Schema Draft 6, "exclusiveMaximum" is a number and can be used instead of "maximum"
            maxValue = isDefined(exclusiveMaximum) ? `${exclusiveMaximum} (exclusive)` : null;
        }
        addFormField("Max Value", maxValue);

        const defaultValue = getValue("default");
        addFormField("Default Value", typeof defaultValue === "object" ? JSON.stringify(defaultValue) : defaultValue);
        let examples = getValue("examples");
        examples = (isDefined(examples) && examples.length > 0) ? examples : null;
        addFormField("Example(s)",
            (examples && typeof examples[0] === "object") ? JSON.stringify(examples) : examples);
        addFormField("Value Pattern", getValue("pattern"));
        addFormField("Value Format", getValue("format"));
        addFormField("Min Length", getValue("minLength"));
        addFormField("Max Length", getValue("maxLength"));
        addFormField("Min Items", getValue("minItems"));
        addFormField("Max Items", getValue("maxItems"));
        addFormField("Items Unique", getValue("uniqueItems") === true ? "Yes" : null);

        return formFields;
    }

    constructor(props) {
        super(props);
        this.renderDefaultSelectionDetails = this.renderDefaultSelectionDetails.bind(this);
    }

    renderDefaultSelectionDetails(parameters) {
        return (
            <div className="jsonschema-inspector-details-content">
                <h3 className="jsonschema-inspector-details-header">Details</h3>
                {this.renderDetailsForm(parameters)}
            </div>
        );
    }

    renderDetailsForm({
        itemSchema, refTargets, columnData, selectionColumnIndex
    }) {
        const formFields = InspectorDetails.collectFormFields(itemSchema, refTargets, columnData, selectionColumnIndex);

        const arrayItems = getFieldValue(itemSchema, "items", refTargets);
        // look-up the kind of value expected in the array (if the schema refers to an array)
        const arrayItemSchema = ((typeof arrayItems === "object") && arrayItems)
            || ((arrayItems === true) && getFieldValue(itemSchema, "additionalItems", refTargets));

        return (
            <div>
                <InspectorDetailsForm fields={formFields} />
                {arrayItemSchema && <hr className="jsonschema-inspector-details-separator" />}
                {arrayItemSchema && <h4 className="jsonschema-inspector-details-header">Array Entry Details</h4>}
                {arrayItemSchema && this.renderDetailsForm({
                    itemSchema: arrayItemSchema,
                    refTargets,
                    columnData,
                    selectionColumnIndex: -1
                })}
            </div>
        );
    }

    render() {
        const {
            columnData, refTargets, renderSelectionDetails, renderEmptyDetails
        } = this.props;
        const lastColumnContainsSelection = columnData.length && columnData[columnData.length - 1].trailingSelection;
        const selectionColumnIndex = columnData.length - (lastColumnContainsSelection ? 1 : 2);
        const trailingSelectionColumn = selectionColumnIndex < 0 ? null : columnData[selectionColumnIndex];
        const itemSchema = trailingSelectionColumn ? trailingSelectionColumn.items[trailingSelectionColumn.selectedItem] : null;
        return (
            <div className="jsonschema-inspector-details">
                {itemSchema
                    && (renderSelectionDetails || this.renderDefaultSelectionDetails)({
                        itemSchema,
                        refTargets,
                        columnData,
                        selectionColumnIndex
                    })}
                {!itemSchema
                    && renderEmptyDetails
                    && renderEmptyDetails({
                        rootColumnSchemas: columnData.length ? columnData[0].items : {}
                    })}
            </div>
        );
    }
}

InspectorDetails.propTypes = {
    columnData: PropTypes.arrayOf(PropTypes.shape({
        items: PropTypes.objectOf(JsonSchemaPropType).isRequired,
        selectedItem: PropTypes.string,
        trailingSelection: PropTypes.bool
    })).isRequired,
    refTargets: PropTypes.objectOf(JsonSchemaPropType).isRequired,
    /** func({ itemSchema: JsonSchema, columnData, refTargets, selectionColumnIndex: number }) */
    renderSelectionDetails: PropTypes.func,
    /** func({ rootColumnSchemas }) */
    renderEmptyDetails: PropTypes.func
};

InspectorDetails.defaultProps = {
    renderSelectionDetails: null,
    renderEmptyDetails: null
};

export default InspectorDetails;
