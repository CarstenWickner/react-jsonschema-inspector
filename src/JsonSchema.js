import RefScope from "./RefScope";
import {
    isNonEmptyObject, listValues, mapObjectValues, mergeObjects
} from "./utils";

class JsonSchema {
    /**
     * Alternative factory method to the constructor, returning "undefined" if the given schema is not a non-empty object.
     *
     * @param {Object} rawSchema the JSON Schema to represent
     * @param {Object} parserConfig configuration affecting how the JSON schema is being traversed/parsed
     * @param {RefScope} scope collection of available $ref targets
     * @returns {JsonSchema|undefined}
     */
    static createIfNotEmpty(rawSchema, parserConfig, scope) {
        return isNonEmptyObject(rawSchema) ? new JsonSchema(rawSchema, parserConfig, scope) : undefined;
    }

    /**
     * Raw JSON Schema
     */
    schema;

    /**
     * Configuration steering how the JSON schema is being traversed/parsed
     */
    parserConfig;

    /**
     * @type RefScope
     */
    scope;

    /**
     * Constructor for a JsonSchema (wrapper).
     *
     * @param {Object} schema the JSON Schema to represent
     * @param {Object} parserConfig configuration affecting how the JSON schema is being traversed/parsed
     * @param {?RefScope} scope collection of available $ref targets
     */
    constructor(schema, parserConfig, scope) {
        this.schema = schema;
        this.parserConfig = parserConfig;
        this.scope = scope || new RefScope(this);
    }

    getRelevantSchemaParts() {
        const { allOf, anyOf, oneOf } = this.schema;
        const includeAnyOf = anyOf && this.parserConfig && this.parserConfig.anyOf === "likeAllOf";
        const includeOneOf = oneOf && this.parserConfig && this.parserConfig.oneOf === "likeAllOf";
        if (allOf || includeAnyOf || includeOneOf) {
            let result = allOf;
            if (includeAnyOf) {
                result = result ? result.concat(anyOf) : anyOf;
            }
            if (includeOneOf) {
                result = result ? result.concat(oneOf) : oneOf;
            }
            return result;
        }
        return undefined;
    }

    /**
     * Look-up of values associated with a specific kind of field in the given JSON schema definition.
     * If the targeted schema is an array, the sub-schemas of its "items" will be ignored.
     *
     * @param {String} fieldName name/key of the field to look-up
     * @param {?Function} mergeFunction optional: how to combine two encountered values into one
     * @param {?Function} mappingFunction optional: converting the raw value before calling the mergeFunction
     * @return {*} result of the given mergeFunction over all encountered values of fields with the given name
     */
    getFieldValue(fieldName, mergeFunction = listValues, mappingFunction) {
        if (!isNonEmptyObject(this.schema)) {
            return undefined;
        }
        if (this.schema.$ref) {
            // schema variable is merely a reference to a separately defined schema
            const referencedSchema = this.scope.find(this.schema.$ref);
            // by convention, if $ref is specified, all other properties in the schema are being ignored
            return referencedSchema.getFieldValue(fieldName, mergeFunction, mappingFunction);
        }
        const rawValue = this.schema[fieldName];
        let value = mappingFunction ? mappingFunction(rawValue, this.parserConfig, this.scope) : rawValue;
        const schemaParts = this.getRelevantSchemaParts();
        if (schemaParts) {
            // schema variable is supposed to be combined with a given list of sub-schemas
            value = schemaParts
                .map(part => new JsonSchema(part, this.parserConfig, this.scope).getFieldValue(fieldName, mergeFunction, mappingFunction))
                .reduce(mergeFunction, value);
        }
        return value;
    }

    /**
     * Version of the getFieldValue() function to look-up of nested JSON schema definitions (while preserving the appropriate reference scope).
     *
     * This method is intended to be used to look-up fields like "items" or "additionalItems".
     *
     * @param {String} fieldName name/key of the field to look-up
     * @param {Function} mergeFunction optional: how to combine two encountered values into one
     * @return {*} result of the given mergeFunction over all encountered values of fields with the given name
     */
    getSchemaFieldValue(fieldName, mergeFunction = listValues) {
        return this.getFieldValue(fieldName, mergeFunction, JsonSchema.createIfNotEmpty);
    }

    /**
     * Determine whether this schema represents an array and if so, return the sub-schema describing its contents.
     *
     * @returns {JsonSchema} type of array contents or null if it is not an array or its content type could not be determined
     */
    getTypeOfArrayItems() {
        return this.getSchemaFieldValue("items") || this.getSchemaFieldValue("additionalItems") || null;
    }

    /**
     * Look-up of any nested sub-schemas that may contain information about "properties" in the given JSON schema definition.
     * If the given JSON schema definition refers is an array, it is assumed to not have relevant information.
     * Instead, the referenced schemas of its "Items" will be returned.
     *
     * This method is intended to be used to look-up fields like "properties" or "required" which are not associated with arrays.
     *
     * BEWARE: if a schema consists of multiple sub-schemas and only one of them defines an "items" field, the other sibling schemas
     * will still be included in the result list - they are just assumed to not have any relevant fields we are looking for.
     *
     * @returns {Array<JsonSchema>} array listing sub-schemas that may define properties about their children
     */
    getPropertyParentSchemas() {
        if (!isNonEmptyObject(this.schema)) {
            return [];
        }
        if (this.schema.$ref) {
            const referencedSchema = this.scope.find(this.schema.$ref);
            // this schema is just a reference to another separately defined schema
            return referencedSchema.getPropertyParentSchemas();
        }
        const schemaParts = this.getRelevantSchemaParts();
        if (schemaParts) {
            let returnedSchemas = [this];
            schemaParts.forEach((rawSchemaPart) => {
                returnedSchemas = returnedSchemas.concat(new JsonSchema(rawSchemaPart, this.parserConfig, this.scope).getPropertyParentSchemas());
            });
            return returnedSchemas;
        }
        if (isNonEmptyObject(this.schema.items)) {
            // unsupported: specifying array of schemas referring to entries at respective positions in described array
            // schema.items contains a single schema, that specifies the type of any value in the described array
            return new JsonSchema(this.schema.items, this.parserConfig, this.scope).getPropertyParentSchemas();
        }
        if (isNonEmptyObject(this.schema.additionalItems)) {
            // the given schema is an array which may specify its content type in "additionalItems"
            return new JsonSchema(this.schema.additionalItems, this.parserConfig, this.scope).getPropertyParentSchemas();
        }
        // if there are no nested schemas, only this schema itself may contain properties
        return [this];
    }

    /**
     * Extract the properties mentioned in this schema â€“ also considering all nested sub-schema by using getPropertyParentSchemas().
     *
     * @returns {Object.<String, JsonSchema>} collection of all properties mentioned in this schema
     */
    getProperties() {
        return this.getPropertyParentSchemas().map(({ schema, scope }) => {
            // properties is an Object.<String, raw-json-schema>
            const { properties, required = [] } = schema;
            const rawProperties = Object.assign(
                {},
                ...required.map(value => ({ [value]: true })),
                isNonEmptyObject(properties) ? properties : {}
            );
            return mapObjectValues(rawProperties, rawSchema => new JsonSchema(rawSchema, this.parserConfig, scope));
        }).reduce(mergeObjects, {});
    }
}

export default JsonSchema;
