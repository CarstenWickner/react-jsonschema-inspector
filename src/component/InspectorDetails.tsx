import PropTypes from "prop-types";
import React, { Component } from "react";

import InspectorDetailsContent from "./InspectorDetailsContent";
import { getColumnDataPropTypeShape } from "./renderDataUtils";
import { RenderColumn, RenderSelectionDetailsFunction, RenderEmptyDetailsFunction, RenderItemsColumn, RenderOptionsColumn } from "../types/Inspector";
import JsonSchemaGroup from "../model/JsonSchemaGroup";

interface InspectorDetailsDefaultProps {
    renderSelectionDetails: RenderSelectionDetailsFunction,
    renderEmptyDetails: RenderEmptyDetailsFunction
};

interface InspectorDetailsProps extends InspectorDetailsDefaultProps {
    columnData: Array<RenderColumn>
};

class InspectorDetails extends Component<InspectorDetailsProps> {
    render() {
        const {
            columnData, renderSelectionDetails, renderEmptyDetails
        } = this.props;
        const lastColumnContainsSelection = columnData.length && columnData[columnData.length - 1].trailingSelection;
        const selectionColumnIndex = columnData.length - (lastColumnContainsSelection ? 1 : 2);
        const trailingSelectionColumn = selectionColumnIndex < 0 ? null : columnData[selectionColumnIndex];
        let itemSchemaGroup: JsonSchemaGroup;
        let optionIndexes: Array<number>;
        if (trailingSelectionColumn) {
            itemSchemaGroup = (trailingSelectionColumn as RenderItemsColumn).items
                ? (trailingSelectionColumn as RenderItemsColumn).items[trailingSelectionColumn.selectedItem as string]
                : (trailingSelectionColumn as RenderOptionsColumn).contextGroup;
            if ((trailingSelectionColumn as RenderOptionsColumn).options) {
                optionIndexes = trailingSelectionColumn.selectedItem as Array<number>;
            }
        }
        return (
            <div className="jsonschema-inspector-details">
                {itemSchemaGroup && renderSelectionDetails && renderSelectionDetails({
                    itemSchemaGroup,
                    selectionColumnIndex,
                    columnData,
                    optionIndexes
                })}
                {itemSchemaGroup && !renderSelectionDetails && (
                    <InspectorDetailsContent
                        itemSchemaGroup={itemSchemaGroup}
                        selectionColumnIndex={selectionColumnIndex}
                        columnData={columnData}
                    />
                )}
                {!itemSchemaGroup && renderEmptyDetails && renderEmptyDetails({
                    rootColumnSchemas: columnData.length ? (columnData[0] as RenderItemsColumn).items : {}
                })}
            </div>
        );
    };

    static propTypes = {
        columnData: PropTypes.arrayOf(PropTypes.shape(getColumnDataPropTypeShape(false))).isRequired,
        /** func({ itemSchema: JsonSchema, selectionColumnIndex: number, columnData }) */
        renderSelectionDetails: PropTypes.func,
        /** func({ rootColumnSchemas }) */
        renderEmptyDetails: PropTypes.func
    };

    static defaultProps: InspectorDetailsDefaultProps = {
        renderSelectionDetails: null,
        renderEmptyDetails: null
    };
}

export default InspectorDetails;
