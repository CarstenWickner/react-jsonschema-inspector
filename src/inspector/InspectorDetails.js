import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';

import JsonSchemaPropType from './JsonSchemaPropType';
import { getPropertyValue } from './utils';

class InspectorDetails extends PureComponent {
    constructor(props) {
        super(props);
        this.renderDefaultSelectionDetails = this.renderDefaultSelectionDetails.bind(this);
    }

    render() {
        const { columnData, renderSelectionDetails, renderEmptyDetails } = this.props;
        const trailingSelectionColumnIndex = columnData.length - (columnData[columnData.length - 1].trailingSelection ? 1 : 2);
        const trailingSelectionColumn = trailingSelectionColumnIndex < 0 ? null : columnData[trailingSelectionColumnIndex];
        const selectedItemSchema = trailingSelectionColumn ? trailingSelectionColumn.items[trailingSelectionColumn.selectedItem] : null;
        return (
            <div className="jsonschema-inspector-details">
                {selectedItemSchema && (renderSelectionDetails || this.renderDefaultSelectionDetails)(selectedItemSchema, columnData, trailingSelectionColumnIndex)}
                {!selectedItemSchema && renderEmptyDetails && renderEmptyDetails(columnData)}
            </div>
        );
    }

    renderDefaultSelectionDetails(itemSchema, columnData, selectionColumnIndex) {
        return (
            <div className="jsonschema-inspector-details-content">
                <h3 className="jsonschema-inspector-details-header">Details</h3>
                {this.renderDetailsForm(itemSchema, columnData, selectionColumnIndex)}
            </div>
        );
    }

    renderDetailsForm(itemSchema, columnData, selectionColumnIndex) {
        const { refTargets } = columnData[0];
        const isRequired = this.isSelectionRequiredInParent(columnData, selectionColumnIndex);

        const title = getPropertyValue(itemSchema, 'title', refTargets);
        const description = getPropertyValue(itemSchema, 'description', refTargets);
        const type = getPropertyValue(itemSchema, 'type', refTargets);
        const constValue = getPropertyValue(itemSchema, 'const', refTargets);
        const enumValue = getPropertyValue(itemSchema, 'enum', refTargets);
        const defaultValue = getPropertyValue(itemSchema, 'default', refTargets);
        const examples = getPropertyValue(itemSchema, 'examples', refTargets);
        const pattern = getPropertyValue(itemSchema, 'pattern', refTargets);
        const format = getPropertyValue(itemSchema, 'format', refTargets);
        const minimum = getPropertyValue(itemSchema, 'minimum', refTargets);
        const exclusiveMinimum = getPropertyValue(itemSchema, 'exclusiveMinimum', refTargets);
        const maximum = getPropertyValue(itemSchema, 'maximum', refTargets);
        const exclusiveMaximum = getPropertyValue(itemSchema, 'exclusiveMaximum', refTargets);
        const minLength = getPropertyValue(itemSchema, 'minLength', refTargets);
        const maxLength = getPropertyValue(itemSchema, 'maxLength', refTargets);
        const minItems = getPropertyValue(itemSchema, 'minItems', refTargets);
        const maxItems = getPropertyValue(itemSchema, 'maxItems', refTargets);
        const uniqueItems = getPropertyValue(itemSchema, 'uniqueItems', refTargets);

        const arrayItems = getPropertyValue(itemSchema, 'items', refTargets);
        // look-up the kind of value expected in the array (if the schema refers to an array)
        const arrayItemSchema = (typeof arrayItems === 'object' && arrayItems) || (typeof arrayItems === 'boolean' && getPropertyValue(itemSchema, 'additionalItems', refTargets));

        return (
            <div>
                <form className="jsonschema-inspector-details-form">
                    {title && this.renderDetailsFormRow("Title", title)}
                    {description && this.renderDetailsFormRow("Description", description)}
                    {isRequired && this.renderDetailsFormRow("Required", "Yes")}
                    {type && this.renderDetailsFormRow("Type", type)}
                    {constValue && this.renderDetailsFormRow("Constant Value", constValue)}
                    {enumValue && this.renderDetailsFormRow("Possible Values", enumValue)}
                    {(defaultValue || defaultValue === false) && this.renderDetailsFormRow("Default Value", typeof defaultValue === 'object' ? JSON.stringify(defaultValue) : defaultValue)}
                    {examples && examples.length && this.renderDetailsFormRow("Example(s)", typeof examples[0] === 'object' ? JSON.stringify(examples) : examples)}
                    {pattern && this.renderDetailsFormRow("Value Pattern", pattern)}
                    {format && this.renderDetailsFormRow("Value Format", format)}
                    {(minimum || minimum === 0) && this.renderDetailsFormRow("Min Value", minimum + (exclusiveMinimum ? " (exclusive)" : " (inclusive)"))}
                    {!minimum && minimum !== 0 && (exclusiveMinimum || exclusiveMinimum === 0) && this.renderDetailsFormRow("Min Value", exclusiveMinimum + " (exclusive)")}
                    {(maximum || maximum === 0) && this.renderDetailsFormRow("Max Value", maximum + (exclusiveMaximum ? " (exclusive)" : " (inclusive)"))}
                    {!maximum && maximum !== 0 && (exclusiveMaximum || exclusiveMaximum === 0) && this.renderDetailsFormRow("Max Value", exclusiveMaximum + " (exclusive)")}
                    {minLength && this.renderDetailsFormRow("Min Length", minLength)}
                    {maxLength && this.renderDetailsFormRow("Max Length", maxLength)}
                    {minItems && this.renderDetailsFormRow("Min Items", minItems)}
                    {maxItems && this.renderDetailsFormRow("Max Items", maxItems)}
                    {typeof uniqueItems === 'boolean' && this.renderDetailsFormRow("Items Unique", uniqueItems ? "Yes" : "No")}
                </form>
                {arrayItemSchema && <hr />}
                {arrayItemSchema && <h4 className="jsonschema-inspector-details-header">Array Entry Details</h4>}
                {arrayItemSchema && this.renderDetailsForm(arrayItemSchema, columnData, -1)}
            </div>
        );
    }

    renderDetailsFormRow(labelText, rowValue) {
        return (
            <div className="jsonschema-inspector-details-form-row" key={labelText}>
                <label className="jsonschema-inspector-details-form-label">{labelText}:</label>
                <span className="jsonschema-inspector-details-form-value">{rowValue.toString()}</span>
            </div>
        );
    }

    isSelectionRequiredInParent(columnData, selectionColumnIndex) {
        if (selectionColumnIndex < 1) {
            // no parent to define any required properties
            return false;
        }
        const { refTargets } = columnData[0];
        const parentColumn = columnData[selectionColumnIndex - 1];
        const parentSchema = parentColumn.items[parentColumn.selectedItem];
        const requiredPropertiesInParent = getPropertyValue(parentSchema, 'required', refTargets) || [];
        const selectionName = columnData[selectionColumnIndex].selectedItem;
        return requiredPropertiesInParent.includes(selectionName);
    }
}

InspectorDetails.propTypes = {
    columnData: PropTypes.arrayOf(PropTypes.shape({
        items: PropTypes.objectOf(JsonSchemaPropType).isRequired,
        selectedItem: PropTypes.string,
        refTargets: PropTypes.objectOf(JsonSchemaPropType)
    })).isRequired,
    renderSelectionDetails: PropTypes.func, // func(selectedItemSchema: JsonSchema, columnData, selectionColumnIndex: number)
    renderEmptyDetails: PropTypes.func // func(columnData)
};

export default InspectorDetails;
