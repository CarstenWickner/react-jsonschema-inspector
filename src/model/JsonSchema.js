import RefScope from "./RefScope";

/**
 * Representation of a Json Schema, offering a number of convenient functions for traversing and extracting information.
 */
export default class JsonSchema {
    /**
     * Raw JSON Schema.
     *
     * @type {object}
     */
    schema;

    /**
     * Configuration steering how the json schema is being traversed/parsed.
     *
     * @type {{oneOf: ?{groupTitle: ?string}, anyOf: ?{groupTitle: ?string}}}
     */
    parserConfig;

    /**
     * @type {RefScope}
     */
    scope;

    /**
     * Constructor for a JsonSchema (wrapper).
     *
     * @param {object} schema - the JSON Schema to represent
     * @param {object} parserConfig - configuration affecting how the json schema is being traversed/parsed
     * @param {?RefScope} scope - collection of available $ref targets (will be generated based on `schema` if not provided)
     */
    constructor(schema, parserConfig, scope) {
        this.schema = schema;
        this.parserConfig = parserConfig;
        this.scope = scope || new RefScope(this);
    }
}
