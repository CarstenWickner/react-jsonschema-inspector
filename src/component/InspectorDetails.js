import PropTypes from "prop-types";
import React from "react";

import InspectorDetailsContent from "./InspectorDetailsContent";

import JsonSchema from "../model/JsonSchema";

const InspectorDetails = ({
    columnData, renderSelectionDetails, renderEmptyDetails
}) => {
    const lastColumnContainsSelection = columnData.length && columnData[columnData.length - 1].trailingSelection;
    const selectionColumnIndex = columnData.length - (lastColumnContainsSelection ? 1 : 2);
    const trailingSelectionColumn = selectionColumnIndex < 0 ? null : columnData[selectionColumnIndex];
    const itemSchema = trailingSelectionColumn ? trailingSelectionColumn.items[trailingSelectionColumn.selectedItem] : null;
    return (
        <div className="jsonschema-inspector-details">
            {itemSchema && renderSelectionDetails && renderSelectionDetails({
                itemSchema,
                selectionColumnIndex,
                columnData
            })}
            {itemSchema && !renderSelectionDetails && (
                <InspectorDetailsContent
                    itemSchema={itemSchema}
                    selectionColumnIndex={selectionColumnIndex}
                    columnData={columnData}
                />
            )}
            {!itemSchema && renderEmptyDetails && renderEmptyDetails({
                rootColumnSchemas: columnData.length ? columnData[0].items : {}
            })}
        </div>
    );
};

InspectorDetails.propTypes = {
    columnData: PropTypes.arrayOf(PropTypes.shape({
        items: PropTypes.objectOf(PropTypes.instanceOf(JsonSchema)).isRequired,
        selectedItem: PropTypes.string,
        trailingSelection: PropTypes.bool
    })).isRequired,
    /** func({ itemSchema: JsonSchema, selectionColumnIndex: number, columnData }) */
    renderSelectionDetails: PropTypes.func,
    /** func({ rootColumnSchemas }) */
    renderEmptyDetails: PropTypes.func
};

InspectorDetails.defaultProps = {
    renderSelectionDetails: null,
    renderEmptyDetails: null
};

export default InspectorDetails;
