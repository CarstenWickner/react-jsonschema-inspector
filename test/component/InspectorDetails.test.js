import React from "react";
import { shallow } from "enzyme";

import InspectorDetails from "../../src/component/InspectorDetails";
import JsonSchema from "../../src/model/JsonSchema";

describe("renders correctly", () => {
    it("with minimal/default props", () => {
        const component = shallow(
            <InspectorDetails
                columnData={[
                    {
                        items: {
                            "Schema One": new JsonSchema({
                                title: "Schema Title",
                                description: "Text"
                            })
                        },
                        selectedItem: "Schema One",
                        trailingSelection: true
                    }
                ]}
            />
        );
        expect(component).toMatchSnapshot();
    });
    it("with empty columnData", () => {
        const component = shallow(
            <InspectorDetails
                columnData={[]}
            />
        );
        expect(component.children().exists()).toBe(false);
    });
    it("with empty columnData and custom renderEmptyDetails", () => {
        const component = shallow(
            <InspectorDetails
                columnData={[]}
                renderEmptyDetails={({ rootColumnSchemas }) => (
                    <span className="custom-empty-details">{Object.keys(rootColumnSchemas).length}</span>
                )}
            />
        );
        expect(component.find(".custom-empty-details").text()).toBe("0");
    });
    it("with no selection", () => {
        const component = shallow(
            <InspectorDetails
                columnData={[
                    {
                        items: {
                            "Schema One": new JsonSchema({
                                title: "Schema Title",
                                description: "Text"
                            })
                        }
                    }
                ]}
            />
        );
        expect(component.children().exists()).toBe(false);
    });
    it("with no selection and custom renderEmptyDetails", () => {
        const component = shallow(
            <InspectorDetails
                columnData={[
                    {
                        items: {
                            "Schema One": new JsonSchema({
                                title: "Schema Title",
                                description: "Text"
                            })
                        }
                    }
                ]}
                renderEmptyDetails={({ rootColumnSchemas }) => (
                    <span className="custom-empty-details">{Object.keys(rootColumnSchemas).length}</span>
                )}
            />
        );
        expect(component.find(".custom-empty-details").text()).toBe("1");
    });
    it("with root array selection", () => {
        const schema = new JsonSchema({
            title: "Schema Title",
            items: { $ref: "#/definitions/itemSchema" },
            definitions: {
                itemSchema: {
                    description: "Array Item",
                    type: "object"
                }
            }
        });
        const columnData = [
            {
                items: {
                    "Schema One": schema
                },
                selectedItem: "Schema One",
                trailingSelection: true
            }
        ];
        const component = shallow(
            <InspectorDetails
                columnData={columnData}
                itemSchema={schema}
                selectionColumnIndex={0}
            />
        );
        const { columnData: contentColumnDataProp, itemSchema, selectionColumnIndex } = component.find("InspectorDetailsContent").props();
        expect(contentColumnDataProp).toEqual(columnData);
        expect(itemSchema).toEqual(schema);
        expect(selectionColumnIndex).toEqual(0);
    });
    it("with root array selection and custom renderSelectionDetails", () => {
        const mainSchema = {
            title: "Schema Title",
            items: { $ref: "#/definitions/itemSchema" },
            definitions: {
                itemSchema: {
                    description: "Array Item",
                    type: "object"
                }
            }
        };
        const columnDataProp = [
            {
                items: {
                    "Schema One": new JsonSchema(mainSchema)
                },
                selectedItem: "Schema One",
                trailingSelection: true
            }
        ];
        const renderSelectionDetails = jest.fn(() => (<span className="custom-selection-details" />));
        const component = shallow(
            <InspectorDetails
                columnData={columnDataProp}
                renderSelectionDetails={renderSelectionDetails}
            />
        );
        expect(component.exists(".custom-selection-details")).toBe(true);
        expect(renderSelectionDetails.mock.calls).toHaveLength(1);
        const {
            itemSchema, columnData, selectionColumnIndex
        } = renderSelectionDetails.mock.calls[0][0];
        expect(itemSchema).toEqual(columnDataProp[0].items["Schema One"]);
        expect(columnData).toEqual(columnDataProp);
        expect(selectionColumnIndex).toBe(0);
    });
});
