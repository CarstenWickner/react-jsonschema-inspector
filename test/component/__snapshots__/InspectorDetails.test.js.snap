// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with minimal/default props 1`] = `
<div
  className="jsonschema-inspector-details"
>
  <InspectorDetailsContent
    columnData={
      Array [
        Object {
          "items": Object {
            "Schema One": JsonSchema {
              "parserConfig": undefined,
              "schema": Object {
                "description": "Text",
                "title": "Schema Title",
              },
              "scope": RefScope {
                "externalRefs": Map {},
                "internalRefs": Map {
                  "#" => [Circular],
                },
                "otherScopes": Array [],
              },
            },
          },
          "selectedItem": "Schema One",
          "trailingSelection": true,
        },
      ]
    }
    itemSchema={
      JsonSchema {
        "parserConfig": undefined,
        "schema": Object {
          "description": "Text",
          "title": "Schema Title",
        },
        "scope": RefScope {
          "externalRefs": Map {},
          "internalRefs": Map {
            "#" => [Circular],
          },
          "otherScopes": Array [],
        },
      }
    }
    selectionColumnIndex={0}
  />
</div>
`;
